synapse/events/spamcheck.py

```diff
@@ -371,25 +441,28 @@ async def user_may_join_room(
                 may_join_room = await delay_cancellation(
                     callback(user_id, room_id, is_invited)
                 )
-            if may_join_room is False:
-                return False
-
-        return True
+            if may_join_room is True or may_join_room is self.NOT_SPAM:
+                continue
+            elif may_join_room is False:
+                return Codes.FORBIDDEN
+            else:
+                return may_join_room
```

Comment:
We should check that the return value is within what's expected (and fallback to e.g. returning `Codes.FORBIDDEN` with an error-level log) so we don't feed Synapse potentially bad data.